src/Control/Carrier.hs

```diff
@@ -1,22 +1,31 @@
 {-# LANGUAGE ConstraintKinds #-}
 module Control.Carrier
-( -- * Re-exports
-  module Control.Carrier.Class
+( -- * Effect requests
+  Has
+, send
+  -- * Re-exports
+, module Control.Carrier.Class
 , module Control.Carrier.Pure
 , module Control.Effect.Class
 , (:+:)(..)
-, Has
-, send
 ) where
 
 import {-# SOURCE #-} Control.Carrier.Class
 import Control.Carrier.Pure
 import Control.Effect.Class
 import Control.Effect.Sum
 
-type Has eff sig m = (Member eff sig, Carrier sig m)
+-- | @m@ is a carrier for @sig@ containing @eff@.
+--
+-- Note that if @eff@ is a sum, it will be decomposed into multiple 'Member' constraints. While this technically allows one to combine multiple unrelated effects into a single 'Has' constraint, doing so has two significant drawbacks:
+--
+-- 1. Due to [a problem with recursive type families](https://gitlab.haskell.org/ghc/ghc/issues/8095), this can lead to significantly slower compiles.
+--
+-- 2. It defeats @ghc@’s warnings for redundant constraints, and thus can lead to a proliferation of redundant constraints as code is changed.
+type Has eff sig m = (Members eff sig, Carrier sig m)
```

Comment:
We’ve had a rather nasty gotcha lurking since #219 that wasn’t discovered until now: Using a `Has` constraint with an effect defined as a sum wouldn’t allow `ghc` to solve `Member` constraints for the members of that sum. Basically, given `Member (A :+: B) sig`, `ghc` wasn’t able to deduce `Member A sig`, which is necessary for `Has A sig m` to be solved.

This remained latent because `NonDet`, the only other effect defined as a sum of other effects, is used exclusively via the `Alternative` interface—we never use `Has NonDet sig m` constraints. It surfaced now because the common pattern is to use `Error e` even when you only need `Throw e`—in no small part because it wasn’t possible to say solely `Throw e` until this PR.

The solution is to decompose the sum into a series of individual `Member` constraints. Note that this technically makes `Has` capable of expressing multiple unrelated effects in a single constraint (synonym):

```haskell
foo :: Has (Reader r :+: State i :+: Error e :+: Writer w) sig m => m ()
```

However, as documented here and on `Members`, this can cause performance problems as well as defeating warnings, so it’s best to avoid it.